### PR TITLE
Added analytics for cask info

### LINKED ISSUE #7566: Casks: Include usage analytics in `brew cask info ...`, just like `brew info ...`
Extend `brew cask info ...` so that it returns usage analytics just like `brew info ...` already returns.

With this recent fix on formulae.brew.sh (https://github.com/Homebrew/formulae.brew.sh/issues/221),  this info can now be fetched with a single query.

For example, `brew cask info` could return information like this:
```
$ brew cask info dropbox
dropbox: latest
https://www.dropbox.com/
Not installed
From: https://github.com/Homebrew/homebrew-cask/blob/master/Casks/dropbox.rb
==> Name
Dropbox
==> Artifacts
Dropbox.app (App)
==> Analytics
install: 2,083 (30 days), 6,551 (90 days), 29,937 (365 days)
```

The analytics returned at the end is my proposed addition.  This just mirrors `brew info`

Usage analytics give users some insight into the popularity of a Cask.  This can information can inform their decision to install it or not.  

I believe all the reasoning that lead to usage analytics being included in `brew info ...` would apply to casks as well.

### PR DESCRIPTION
**Pull request by tharun208**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This PR contains that add analytics info to the `brew cask info` command.

Fixes #7566

### CODE REVIEW

File: Library/Homebrew/cask/cmd/info.rb [hunk 1]
```diff
@@ -41,11 +41,12 @@ def self.get_info(cask)
         output << artifact_info(cask) + "\n"
         caveats = Installer.caveats(cask)
         output << caveats if caveats
-        output
+        puts output
+        Analytics.formula_output(cask)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cmd/info.rb:45]: Could this method get renamed from `formula_output` (as this isn't a formula output here).

File: Library/Homebrew/cask/cmd/info.rb [hunk 2]
```diff
@@ -41,11 +41,12 @@ def self.get_info(cask)
         output << artifact_info(cask) + "\n"
         caveats = Installer.caveats(cask)
         output << caveats if caveats
-        output
+        puts output
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cmd/info.rb:44]: This needs to not `puts` as it's used in `Library/Homebrew/extend/os/Mac/missing_formula.rb`

File: Library/Homebrew/test/missing_formula_spec.rb [hunk 3]
```diff
@@ -105,7 +105,7 @@
       let(:formula) { "local-caffeine" }
       let(:show_info) { true }
 
-      it { is_expected.to match(/Found a cask named "local-caffeine" instead.\n\nlocal-caffeine: 1.2.3\n/) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/missing_formula_spec.rb:108]: This test shouldn't be changed.

File: Library/Homebrew/test/missing_formula_spec.rb [hunk 4]
```diff
@@ -148,7 +148,6 @@
       let(:command) { "info" }
 
       it { is_expected.to match(/Found a cask named "local-caffeine" instead./) }
-      it { is_expected.to match(/local-caffeine: 1.2.3/) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/missing_formula_spec.rb:151]: This test shouldn't be changed.

File: Library/Homebrew/cask/cmd/info.rb (hunk 1 above)
Comment (tharun208) [reply Library/Homebrew/cask/cmd/info.rb]: done

**Comment by tharun208** [comment]
@MikeMcQuaid I am just thinking we should need to refractor the `can run be run with a url twice` 
in cask tests.Since it is using a real cask, the test will fail if analytics data changes.

**Comment by varenc** [comment]
@tharun208 yea seems like those live analytics numbers will break the test when they change.  The equivalent [`brew info` test](https://github.com/Homebrew/brew/blob/540d46d1bf41f17d7c9f34f92ced566a98399e76/Library/Homebrew/test/cmd/info_spec.rb#L13) just [creates a formula](https://github.com/Homebrew/brew/blob/540d46d1bf41f17d7c9f34f92ced566a98399e76/Library/Homebrew/test/cmd/info_spec.rb#L13) for testing on the fly, but it also uses `--json` which doesn't include the analytics info anyway.   

One minimal/cleanish approach could be to change the tests so that on the Analytics portion they don't check for an exact string match and just check that an Analytics section exists instead.  Though that'll still probably be more code than the new feature itself.

(I'm just a rando watching from the sidelines so don't put too much weight on my opinions.  I filed that #7566 issue and was too slow to get started on the PR! Psyched this feature is on the way)

**Comment by MikeMcQuaid** [comment]
I would suggest mocking the `cask_api_json` call e.g. `expect(...).to receive(:...).and_return(...)` so that it returns the data you specify.

**Comment by tharun208** [comment]
> @tharun208 yea seems like those live analytics numbers will break the test when they change. The equivalent [`brew info` test](https://github.com/Homebrew/brew/blob/540d46d1bf41f17d7c9f34f92ced566a98399e76/Library/Homebrew/test/cmd/info_spec.rb#L13) just [creates a formula](https://github.com/Homebrew/brew/blob/540d46d1bf41f17d7c9f34f92ced566a98399e76/Library/Homebrew/test/cmd/info_spec.rb#L13) for testing on the fly, but it also uses `--json` which doesn't include the analytics info anyway.
> 
> One minimal/cleanish approach could be to change the tests so that on the Analytics portion they don't check for an exact string match and just check that an Analytics section exists instead. Though that'll still probably be more code than the new feature itself.
> 
> (I'm just a rando watching from the sidelines so don't put too much weight on my opinions. I filed that #7566 issue and was too slow to get started on the PR! Psyched this feature is on the way)

@varenc , that is what I am checking with mike and I thought of mocking the cask api

**Comment by tharun208** [comment]
> I would suggest mocking the `cask_api_json` call e.g. `expect(...).to receive(:...).and_return(...)` so that it returns the data you specify.

@MikeMcQuaid , sure will make the changes

**Comment by tharun208** [comment]
> > I would suggest mocking the `cask_api_json` call e.g. `expect(...).to receive(:...).and_return(...)` so that it returns the data you specify.
> 
> @MikeMcQuaid , sure will make the changes, And can you help me by adding the mock and I got one issue - all the tests are always passing in my local. Even though, there is a failing test.

**Comment by MikeMcQuaid** [comment]
@tharun208 It sounds like you need to use `brew tests --online` and/or edit the files in `brew --repo` rather than your local checkout.

**Comment by tharun208** [comment]
@MikeMcQuaid , Can you help me with mocking the cask API  to get analytics. I tried different steps, but not working

**Comment by MikeMcQuaid** [comment]
@tharun208 I can but you need to ask more specific questions and/or push what you've tried to this PR.

**Comment by tharun208** [comment]
> @tharun208 I can but you need to ask more specific questions and/or push what you've tried to this PR.

Basically, I am reusing the formulae API endpoint to get the cask's analytics and showing that result in the `info` command.

**Comment by tharun208** [comment]
> > @tharun208 I can but you need to ask more specific questions and/or push what you've tried to this PR.
> 
> Basically, I am reusing the formulae API endpoint to get the cask's analytics and showing that result in the `info` command. I already have done the implementation. but I am trying to fix the tests with mocking the analytics data.

**Comment by MikeMcQuaid** [comment]
> Basically, I am reusing the formulae API endpoint to get the cask's analytics and showing that result in the `info` command.

Sorry to be pedantic but that's still not an actual question I can help you with.

If you want a maintainer to help you with something you need to explain:
- What you were trying to do (and why)
- What happened
- What you expected to happen
- Step-by-step reproduction instructions (with as minimal input data as possible)

In this case probably the best thing is to push whatever code you tried locally and we can talk about that code. I can't help fix code I can't see.

**Comment by tharun208** [comment]
Since I am adding analytics data to the `info command`. I am refactoring the test in [cask info tests](https://github.com/Homebrew/brew/blob/6e543f7289636a83feb4976c6359a88ace267b85/Library/Homebrew/test/cask/cmd/info_spec.rb) by mocking the analytics data why because it is analytics data and will change frequently.
So, I tried using mocking the `formulae_api_json` to return static analytics data. But mocking is not working as expected. I explicitly tried to fail the test by giving wrong arguments to the function, But is throwing `ArgumentError:
       wrong number of arguments (given 0, expected 1)`.

File: Library/Homebrew/test/cask/cmd/info_spec.rb [hunk 5]
```diff
@@ -141,6 +142,24 @@
...
+          },
+          90d: {
+            docker: 2000
+          },
+          365d: {
+            docker: 3000
+          }
+        }
+      }
+    }")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cask/cmd/info_spec.rb:162]: You need something like this (untested so you'll need to test locally):

```suggestion
  it "can run be run with a url twice", :needs_network do
    expect(Utils::Analytics).to receive(:formulae_api_json).with("docker").and_return("{
      analytics: {
        install: {
          30d: {
            docker: 1000
          },
          90d: {
            docker: 2000
          },
          365d: {
            docker: 3000
          }
        }
      }
    }")
    expect {
```

File: Library/Homebrew/test/cask/cmd/info_spec.rb [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 require_relative "shared_examples/requires_cask_token"
 require_relative "shared_examples/invalid_option"
+require_relative "../../../utils"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cask/cmd/info_spec.rb:5]:
```suggestion
require "utils"
```

File: Library/Homebrew/utils/analytics.rb [hunk 7]
```diff
@@ -179,6 +176,20 @@ def formula_output(f)
         end
       end
 
+      def formula_output(f)
+        json = formulae_api_json("#{formula_path}/#{f}.json")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/analytics.rb:180]: May want to rename `formulae_api_json` to `formulae_brew_sh_json`

File: Library/Homebrew/test/cask/cmd/info_spec.rb [hunk 8]
```diff
@@ -155,6 +173,8 @@
       Docker CE
       ==> Artifacts
       Docker.app (App)
+      ==> Analytics
+      install: 16,997 (30 days), 54,056 (90 days), 239,482 (365 days)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cask/cmd/info_spec.rb:177]: There numbers seem like they should match those in the `expect` call?

Comment (tharun208) [reply Library/Homebrew/test/cask/cmd/info_spec.rb:177]: i tried with above, the mocking is happening but it cant able to parse information from the mocked data.

Comment (tharun208) [reply Library/Homebrew/test/cask/cmd/info_spec.rb:177]: sorry, forgot to parse the data :)

File: Library/Homebrew/test/cask/cmd/info_spec.rb [hunk 9]
```diff
@@ -140,6 +142,9 @@
   end
 
   it "can run be run with a url twice", :needs_network do
+    analytics = '{"analytics":{"install":{"30d":{"docker": 1000},"90d":{"docker": 2000},"365d":{"docker": 3000}}}}'
+    expect(Utils::Analytics).to receive(:formulae_brew_sh_json).twice.with("cask/docker.json")
+    .and_return(JSON.parse(analytics))
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cask/cmd/info_spec.rb:147]: Instead of `JSON.parse` consider using a Ruby dictionary instead.

Comment (tharun208) [reply Library/Homebrew/test/cask/cmd/info_spec.rb]: done!

File: Library/Homebrew/test/cask/cmd/info_spec.rb [hunk 10]
```diff
@@ -140,6 +141,15 @@
   end
 
   it "can run be run with a url twice", :needs_network do
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cask/cmd/info_spec.rb:143]:
```suggestion
  it "can run be run with a url twice and returns analytics", :needs_network do
```

File: Library/Homebrew/cask/cmd/info.rb [hunk 11]
```diff
@@ -46,6 +46,7 @@ def self.get_info(cask)
 
       def self.info(cask)
         puts get_info(cask)
+        Analytics.cask_output(cask)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cmd/info.rb:49]:
```suggestion
        Utils::Analytics.cask_output(cask)
```

and perhaps just require in this file?

File: Library/Homebrew/cask/cmd/abstract_command.rb [hunk 12]
```diff
@@ -2,12 +2,14 @@
...
 require_relative "options"
 require "search"
+require "utils"
 
 module Cask
   class Cmd
     class AbstractCommand
       include Options
       include Homebrew::Search
+      include Utils
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cmd/abstract_command.rb:12]:
```suggestion
```

File: Library/Homebrew/cask/cmd/info.rb (hunk 11 above)
Comment (tharun208) [reply Library/Homebrew/cask/cmd/info.rb:49]: I tried this and it is not working. Since `Info` is a class

Comment (MikeMcQuaid) [reply Library/Homebrew/cask/cmd/info.rb:49]: Might need to be `::Utils::Analytics`?

Comment (tharun208) [reply Library/Homebrew/cask/cmd/info.rb:49]: yes

**Comment by MikeMcQuaid** [comment]
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @tharun208!

**Comment by tharun208** [comment]
@MikeMcQuaid, Thanks for helping :) and would like to contribute more

**Comment by varenc** [comment]
@tharun208 I send you my thanks as well! I've been wanting this feature for awhile. I even made an alias for quickly fetching cask analytics using the HTTP API and jq...very happy I can toss that out and just have that in directly in homebrew now. Cheers!

**Pull request merged**